### PR TITLE
Refactor TraceSource -> Microsoft.Extensions.Logging

### PR DESCRIPTION
**Pull request by danystem**
Hi @NicolasDorier,

Just wanted to get the PR open - so you have visibility of it going in the right direction and to kick of the CI?

I have current left all the Trace stuff in place commented out - but will obviously clean up when refined.

Let me know your thoughts.

Dan

### CODE REVIEW

**Comment by danystem** [comment]
Hmm - do we still need to support netstandard1.1 as a minimum?

**Comment by NicolasDorier** [comment]
you can remove the correlation stuff

**Comment by NicolasDorier** [comment]
Don't comment out stuff, remove stuff

**Comment by danystem** [comment]
Absolutely i will, just wanted to get the first cut in there and run it through the ci. Only travis failing now, unsure why currently.

**Comment by D3m0nKingx** [comment]
@danystem might be this:

```C#
[39;49m[31m[xUnit.net 00:02:21.1773980]     NBitcoin.Tests.RPCClientTests.CanGetPrivateKeysFromLockedAccount [FAIL]

[39;49m[31mFailed   NBitcoin.Tests.RPCClientTests.CanGetPrivateKeysFromLockedAccount

[39;49m[39;49m[31mError Message:

[39;49m[39;49m[31m System.Net.WebException : An error occurred while sending the request. The server returned an invalid or unrecognized response.

---- System.Net.Http.HttpRequestException : An error occurred while sending the request.

-------- System.IO.IOException : The server returned an invalid or unrecognized response.

[39;49m[39;49m[31mStack Trace:

[39;49m[39;49m[31m   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult)

   at System.Net.WebRequest.<>c.<GetResponseAsync>b__68_2(IAsyncResult iar)

   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)

--- End of stack trace from previous location where exception was thrown ---

   at NBitcoin.RPC.RPCClient.SendCommandAsyncCore(RPCRequest request, Boolean throwIfRPCError) in /Users/travis/build/MetacoSA/NBitcoin/NBitcoin/RPC/RPCClient.cs:line 704

   at NBitcoin.RPC.RPCClient.SendCommandAsyncCore(RPCRequest request, Boolean throwIfRPCError) in /Users/travis/build/MetacoSA/NBitcoin/NBitcoin/RPC/RPCClient.cs:line 714

   at NBitcoin.RPC.RPCClient.SendCommandAsync(RPCRequest request, Boolean throwIfRPCError) in /Users/travis/build/MetacoSA/NBitcoin/NBitcoin/RPC/RPCClient.cs:line 649

   at NBitcoin.RPC.RPCClient.SendCommand(RPCRequest request, Boolean throwIfRPCError) in /Users/travis/build/MetacoSA/NBitcoin/NBitcoin/RPC/RPCClient.cs:line 462

   at NBitcoin.RPC.RPCClient.SendCommand(RPCOperations commandName, Object[] parameters) in /Users/travis/build/MetacoSA/NBitcoin/NBitcoin/RPC/RPCClient.cs:line 378

   at NBitcoin.Tests.RPCClientTests.CanGetPrivateKeysFromLockedAccount() in /Users/travis/build/MetacoSA/NBitcoin/NBitcoin.Tests/RPCClientTests.cs:line 747

----- Inner Stack Trace -----

   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)

   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)

   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)

   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)

   at System.Net.HttpWebRequest.SendRequest()

   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult)

----- Inner Stack Trace -----

   at System.Net.Http.HttpConnection.FillAsync()

   at System.Net.Http.HttpConnection.ReadNextResponseHeaderLineAsync(Boolean foldedHeadersAllowed)

   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
```

**Comment by NicolasDorier** [comment]
Remove stuff commented out

File: NBitcoin/Crypto/ECDSASignature.cs [hunk 1]
```diff
@@ -139,7 +141,7 @@ public static bool IsValidDER(byte[] bytes)
 			}
 			catch(Exception ex)
 			{
-				Utils.error("Unexpected exception in ECDSASignature.IsValidDER " + ex.Message);
+				//Logs.Utils.LogError(default, ex, "Unexpected exception in ECDSASignature.IsValidDER"); //Potential Side Effect
```
Comment (knocte) [inline NBitcoin/Crypto/ECDSASignature.cs]: if you remove this log call here then the catch is a silent-swallower of errors.

Comment (NicolasDorier) [reply NBitcoin/Crypto/ECDSASignature.cs]: yeah, I think this is fine, anyway nobody ever could read those logs.

Comment (knocte) [reply NBitcoin/Crypto/ECDSASignature.cs]: what do you mean with nobody could read these logs?

File: NBitcoin/Protocol/NodeServer.cs [hunk 2]
```diff
@@ -178,43 +177,45 @@ private void EndAccept(SocketAsyncEventArgs args)
...
 						return;
-					NodeServerTrace.Information("Client connection accepted : " + client.RemoteEndPoint);
-					using(var cancel = CancellationTokenSource.CreateLinkedTokenSource(_Cancel.Token))
-					{
-						cancel.CancelAfter(TimeSpan.FromSeconds(10));
 
-						var stream = new NetworkStream(client, false);
-						while(true)
+					Logs.NodeServer.LogInformation("Client connection accepted {remoteEndPoint}", client.RemoteEndPoint);
+					var cancel = CancellationTokenSource.CreateLinkedTokenSource(_Cancel.Token);
```
Comment (NicolasDorier) [inline NBitcoin/Protocol/NodeServer.cs]: why you removed the using?

File: NBitcoin/Protocol/NodeServer.cs [hunk 3]
```diff
@@ -266,27 +267,22 @@ internal void ExternalAddressDetected(IPAddress iPAddress)
 		{
 			if(!ExternalEndpoint.Address.IsRoutable(AllowLocalPeers) && iPAddress.IsRoutable(AllowLocalPeers))
 			{
-				NodeServerTrace.Information("New externalAddress detected " + iPAddress);
+				Logs.NodeServer.LogInformation("New externalAddress detected {externalAddress}", iPAddress);
```
Comment (NicolasDorier) [inline NBitcoin/Protocol/NodeServer.cs]: You forgot the $ sign in front of the string

File: NBitcoin/Protocol/Message.cs [hunk 4]
```diff
@@ -131,8 +132,11 @@ public void ReadWrite(BitcoinStream stream)
 
 					var payloadType = PayloadAttribute.GetCommandType(Command);
 					var unknown = payloadType == typeof(UnknowPayload);
-					if(unknown)
-						NodeServerTrace.Trace.TraceEvent(TraceEventType.Warning, 0, "Unknown command received : " + Command);
+					if (unknown)
+					{
```
Comment (NicolasDorier) [inline NBitcoin/Protocol/Message.cs]: remove braces

File: NBitcoin/Protocol/AddressManager.cs [hunk 5]
```diff
@@ -1208,7 +1212,7 @@ internal void DiscoverPeers(Network network, NodeConnectionParameters parameters
 							if(found >= peerToFind)
 								peerTableFull.Cancel();
 							else
-								NodeServerTrace.Information("Need " + (-found + peerToFind) + " more peers");
+								Logs.NodeServer.LogInformation("Need {neededPeerCount} more peers", (-found + peerToFind));
```
Comment (NicolasDorier) [inline NBitcoin/Protocol/AddressManager.cs]: you forgot the $ sign

**Comment by NicolasDorier** [comment]
Remove structured logging, I don't think it will ever get used. (where you don't use the $ sign)

File: NBitcoin/Crypto/ECDSASignature.cs (hunk 1 above)
Comment (NicolasDorier) [reply NBitcoin/Crypto/ECDSASignature.cs]: Utils.error was doing nothing. And if it did, this specific error message would not help anyone.

Comment (danystem) [reply NBitcoin/Crypto/ECDSASignature.cs]: Do you want to retain logging here then or not? Reason it was commented out is that this causes the unit tests to fail if left in - a weird behavior. I'll look at resolving but not worth it if silent exception is acceptable?

Comment (knocte) [reply NBitcoin/Crypto/ECDSASignature.cs]: IMO if Utils.error was doing nothing, let's make it do something in this PR instead of removing it altogether

Comment (knocte) [reply NBitcoin/Crypto/ECDSASignature.cs]: or just rethrow

**Comment by danystem** [comment]
> Remove structured logging, I don't think it will ever get used. (where you don't use the $ sign)

@NicolasDorier do you not think it would be useful to keep the structured logging. If we can start padding out some of the log messages with additional meta data it means we can filter easily on these. For example we could log at all the logging around a certain peer etc? 

This would definitely be very useful to myself and as ILogger is abstracted appropriately to support this, does it lose us anything by not doing it?

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-2.1#log-message-template

File: NBitcoin/Protocol/Message.cs (hunk 4 above)
Comment (danystem) [reply NBitcoin/Protocol/Message.cs]: Removed now - apologies.

File: NBitcoin/Crypto/ECDSASignature.cs (hunk 1 above)
Comment (danystem) [reply NBitcoin/Crypto/ECDSASignature.cs]: @NicolasDorier await your steer?

**Comment by NicolasDorier** [comment]
Ok, if you feel strongly for it.

**Comment by NicolasDorier** [comment]
Remove the stuff commented out, and it seems good to me.

**Comment by danystem** [comment]
Should be good now

File: NBitcoin.Tests/ProtocolTests.cs [hunk 6]
```diff
@@ -132,6 +135,13 @@ internal void AddDisposable(IDisposable disposable)
 	}
 	public class ProtocolTests
 	{
+
+		//public ProtocolTests(ITestOutputHelper testOutputHelper)
+		//{
+		//	var logger = new XunitLogger(testOutputHelper, "ProtocolTests");
+		//	Logs.Configure(new FuncLoggerFactory(l=>logger));
+		//}
```
Comment (knocte) [inline NBitcoin.Tests/ProtocolTests.cs]: still commented code here

File: NBitcoin.Tests/RPCClientTests.cs [hunk 7]
```diff
@@ -22,6 +25,13 @@ namespace NBitcoin.Tests
 	public class RPCClientTests
 	{
 		const string TestAccount = "NBitcoin.RPCClientTests";
+
+		//public RPCClientTests(ITestOutputHelper testOutputHelper)
+		//{
+		//	var logger = new XunitLogger(testOutputHelper, "RPCClientTests");
+		//	Logs.Configure(new FuncLoggerFactory(l=>logger));
+		//}
+
```
Comment (knocte) [inline NBitcoin.Tests/RPCClientTests.cs]: and here

File: NBitcoin/Protocol/Node.cs [hunk 8]
```diff
@@ -4,7 +4,7 @@
 using System;
 using System.Collections.Concurrent;
 using System.Collections.Generic;
-using System.Diagnostics;
+//using System.Diagnostics; TODO:Remove
```
Comment (knocte) [inline NBitcoin/Protocol/Node.cs]: commented code here

File: NBitcoin/Protocol/Node.cs [hunk 9]
```diff
@@ -173,18 +175,18 @@ public CancellationTokenSource Cancel
...
+//#else
+//			public
+//#endif
+// TraceCorrelation TraceCorrelation
+//			{
+//				get
+//				{
+//					return Node.TraceCorrelation;
+//				}
+//			}
```
Comment (knocte) [inline NBitcoin/Protocol/Node.cs]: and here

**Comment by knocte** [comment]
Still lots of places with commented code.

**Comment by danystem** [comment]
Sorry bad rebase - will resolve this all now....

File: NBitcoin.Tests/ProtocolTests.cs (hunk 6 above)
Comment (danystem) [reply NBitcoin.Tests/ProtocolTests.cs]: fixed now

File: NBitcoin.Tests/RPCClientTests.cs (hunk 7 above)
Comment (danystem) [reply NBitcoin.Tests/RPCClientTests.cs]: fixed now

File: NBitcoin/Protocol/Node.cs (hunk 9 above)
Comment (danystem) [reply NBitcoin/Protocol/Node.cs]: fixed now

File: NBitcoin/Protocol/Node.cs (hunk 8 above)
Comment (danystem) [reply NBitcoin/Protocol/Node.cs]: fixed now

**Comment by danystem** [comment]
Should be all good now @knocte

File: NBitcoin/Protocol/NodeServer.cs [hunk 10]
```diff
@@ -168,72 +170,69 @@ private void Accept_Completed(object sender, SocketAsyncEventArgs e)
...
+			try
 			{
-				Socket client = null;
-				try
+				if(args.SocketError != SocketError.Success)
+					throw new SocketException((int)args.SocketError);
+				client = args.AcceptSocket;
+				if(_Cancel.IsCancellationRequested)
+					return;
+			    Logs.NodeServer.LogInformation("Client connection accepted {remoteEndPoint}", client.RemoteEndPoint);
```
Comment (knocte) [inline NBitcoin/Protocol/NodeServer.cs]: mixed tabs+spaces in this line

File: NBitcoin/Protocol/NodeServer.cs [hunk 11]
```diff
@@ -168,72 +170,69 @@ private void Accept_Completed(object sender, SocketAsyncEventArgs e)
...
+						{
+							Socket = client,
+							Message = message,
+							Length = counter.ReadenBytes,
+							Node = null,
+						});
+						if(message.Payload is VersionPayload)
+							break;
+						else
+						    Logs.NodeServer.LogError("The first message of the remote peer did not contained a Version payload");
```
Comment (knocte) [inline NBitcoin/Protocol/NodeServer.cs]: mixed tabs+spaces in this other line

File: NBitcoin/Protocol/NodeServer.cs (hunk 3 above)
Comment (knocte) [reply NBitcoin/Protocol/NodeServer.cs]: this is still an issue ^

File: Stubs/TraceStub/TraceSource.cs [hunk 12]
```diff
@@ -1,123 +1,123 @@
...
-			return CreateTraceSourceFactory(name);
-		}
+//namespace System.Diagnostics
+//{
+//	public class TraceSourceFactory
+//	{
+//		public static TraceSource CreateTraceSource(string name)
+//		{
+//			return CreateTraceSourceFactory(name);
+//		}
```
Comment (knocte) [inline Stubs/TraceStub/TraceSource.cs]: still commented code here

File: Stubs/TraceStub/TraceSource.cs [hunk 13]
```diff
@@ -1,123 +1,123 @@
...
+//{
+//	public class TraceSourceFactory
+//	{
+//		public static TraceSource CreateTraceSource(string name)
+//		{
+//			return new TraceSource(name);
+//		}
+//	}
+//}
+//#endif
```
Comment (knocte) [inline Stubs/TraceStub/TraceSource.cs]: still commented code in this file

File: NBitcoin/Protocol/NodeServer.cs (hunk 3 above)
Comment (NicolasDorier) [reply NBitcoin/Protocol/NodeServer.cs]: no, actually he wants to use structured logging

**Pull request merged**